events/app.py

```diff
@@ -88,40 +89,56 @@ def get_one_event(event_id):
     pass
 
 
-class Event(object):
-    attributes = [
-        'event_id',
-        'name',
-        'description',
-        'author',
-        'created_at',
-        'event_time']
+attributes = [
+    'event_id',
+    'name',
+    'description',
+    'author',
+    'created_at',
+    'event_time']
+
+EventTuple = namedtuple('EventTuple', attributes)
+
 
+class Event(object):
     def __init__(self, info):
-        self.info = dict(info)
+        """Constructs an event object represented by a EventTuple namedtuple.
+
+        Creating the EventTuple raises a ValueError if any attributes are
+        missing or if extra attributes are included.
+        """
         # event_id is defined by the database, not the user
-        if "event_id" not in self.info:
-            self.info['event_id'] = None
+        if "event_id" not in info:
+            info['event_id'] = None
+        if '_id' in info:       # found DB-generated ID, override given one
+            info['event_id'] = info['_id']
+            del info['_id']
+        self.info = EventTuple(**info)
 
     def __eq__(self, other):
         """Determines if two events have the same info, excluding event_id."""
-        for att in self.attributes:
-            if att != 'event_id' and (self.info[att] != other.info[att]):
-                return False
-        return True
-
-    def is_valid(self):
-        """Checks that all of the required attributes are in the event info."""
-        for att in self.attributes:
-            if att not in self.info:
+        if not isinstance(other, Event):
+            return False
+        for att in attributes:
+            if att != 'event_id' and (
+                    self.get_info()[att] != other.get_info()[att]):
                 return False
+            if att == 'event_time' or att == 'created_at':
+                pass
         return True
 
     def add_to_db(self, events_collection):
         """Adds the event to the specified collection."""
-        events_collection.insert_one(self.info)
+        events_collection.insert_one(self.get_info())
         return self.info
 
+    def get_info(self):
```

Comment:
Consider using @property here to make this a read property, rather than as a method.

In particular, if you make Event hold all the data directly, then the "info as dict" version could be called `.info` or `.dict` or the like.

Reply:
Done in [10703b5](https://github.com/knative-portability/large-events/pull/4/commits/10703b581ed58e4bcb9159704e48834f90275c93)